utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java

```diff
@@ -332,97 +324,91 @@ private void handleMapViewClicked(MouseEvent e) {
-
-              // check the utility network source type of every utility element
-              utilityElementList.forEach(utilityElement -> {
-
-                // check if the network source is a junction or an edge
-                if (utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.JUNCTION) {
-
-                  // check if the feature has a terminal configuration and multiple terminals
-                  if (utilityElement.getAssetType().getTerminalConfiguration() != null) {
-                    UtilityTerminalConfiguration utilityTerminalConfiguration = utilityElement.getAssetType().getTerminalConfiguration();
-                    List<UtilityTerminal> terminals = utilityTerminalConfiguration.getTerminals();
-
-                    if (terminals.size() > 1) {
-                      // prompt the user to select a terminal for this feature
-                      Optional<UtilityTerminal> userSelectedTerminal = promptForTerminalSelection(terminals);
-
-                      // apply the selected terminal
-                      if (userSelectedTerminal.isPresent()) {
-                        UtilityTerminal terminal = userSelectedTerminal.get();
-                        utilityElement.setTerminal(terminal);
-                        // show the terminals name in the status label
-                        String terminalName = terminal.getName() != null ? terminal.getName() : "default";
-                        statusLabel.setText("Feature added at terminal: " + terminalName);
-
-                        // don't create the element if no terminal was selected
-                      } else {
-                        statusLabel.setText("No terminal selected - no feature added");
+        mapView.identifyLayersAsync(screenPoint, 10, false).toCompletableFuture().whenComplete(
+          (identifyLayerResults, ex) -> {
+            if (ex == null) {
+
+              // return if no features are identified
+              if (!identifyLayerResults.isEmpty()) {
+
+                // get and store a list of features from the result of the query (there may be more than one)
+                List<ArcGISFeature> listOfFeatures = new ArrayList<>();
+                identifyLayerResults.forEach(result -> listOfFeatures.add((ArcGISFeature) result.getElements().get(0)));
+
+                // create utility element for each feature and store it in a list
+                List<UtilityElement> utilityElementList = new ArrayList<>();
+                listOfFeatures.forEach(feature -> utilityElementList.add(utilityNetwork.createElement(feature)));
+
+                // check the utility network source type of every utility element
+                utilityElementList.forEach(utilityElement -> {
+
+                  // check if the network source is a junction or an edge
+                  if (utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.JUNCTION) {
```

Comment:
We've got a decent bit of if/else-if logic so unless I'm missing something this only really works as two separate stream operations - tried that and it's not better than what we already have to my eyes.

Reply:
This is what I tried:
```java
                 // handle junctions
                  utilityElementList.stream().filter(utilityElement ->
                          utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.JUNCTION).toList()
                      .stream().filter(junction -> junction.getAssetType().getTerminalConfiguration() != null && junction.getAssetType().getTerminalConfiguration().getTerminals().size() > 1).toList()
                      .forEach(junction -> {
                        // prompt the user to select a terminal for this feature
                        Optional<UtilityTerminal> userSelectedTerminal = promptForTerminalSelection(junction.getAssetType().getTerminalConfiguration().getTerminals());

                        // apply the selected terminal
                        if (userSelectedTerminal.isPresent()) {
                          UtilityTerminal terminal = userSelectedTerminal.get();
                          junction.setTerminal(terminal);
                          // show the terminals name in the status label
                          String terminalName = terminal.getName() != null ? terminal.getName() : "default";
                          statusLabel.setText("Feature added at terminal: " + terminalName);

                          // don't create the element if no terminal was selected
                        } else {
                          statusLabel.setText("No terminal selected - no feature added");
                        }
                      });

                  // handle edges
                  utilityElementList.stream().filter(utilityElement -> utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.EDGE).toList()
                      .forEach(edge -> {
                        // get the geometry of the identified feature as a polyline, and remove the z component
                        Polyline polyline = (Polyline) GeometryEngine.removeZ(listOfFeatures.get(0).getGeometry());

                        // compute how far the clicked location is along the edge feature
                        double fractionAlongEdge = GeometryEngine.fractionAlong(polyline, mapPoint, -1);
                        if (Double.isNaN(fractionAlongEdge)) {
                          new Alert(Alert.AlertType.ERROR, "Cannot add starting location / barrier here.");
                          return;
                        }

                        // set the fraction along edge
                        edge.setFractionAlongEdge(fractionAlongEdge);
                        // update the status label text
                        statusLabel.setText("Fraction along edge: " + Math.round(edge.getFractionAlongEdge() * 1000d) / 1000d);
                      });
```
I think I agree in terms of legibility preferring an if/else, however if we are better making heavier use of Streams like this then I think we should. We can work on improving the comments to better match what's going on to accompany the chanining. 

Thoughts @colinanderson on the above approach and if we should a) go ahead and implement for this sample and b) make better/more use of streams throughout the samples?

Reply:
If I was doing it I'd probably have used the filter just because I prefer that to looping and branching but it does mean sometimes you end up streaming twice (or more). Either way is fine.